truffle/src/main/java/org/jruby/truffle/nodes/core/RopeNodes.java

```diff
@@ -0,0 +1,438 @@
+            final int characterLength = StringSupport.unpackResult(packedLengthAndCodeRange);
+
+            /*
+            if (base.depth() >= 10) {
+                System.out.println("SubstringRope depth: " + (base.depth() + 1));
+            }
+            */
+
+            return new SubstringRope(base, offset, byteLength, characterLength, codeRange);
+        }
+
+        protected static boolean sameAsBase(Rope base, int offset, int byteLength) {
+            return (byteLength - offset) == base.byteLength();
+        }
+
+    }
+
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "left"),
+            @NodeChild(type = RubyNode.class, value = "right"),
+            @NodeChild(type = RubyNode.class, value = "encoding")
+    })
+    public abstract static class MakeConcatNode extends RubyNode {
+
+        protected static final int SHORT_LEAF_BYTESIZE_THRESHOLD = 128;
+
+        @Child private MakeLeafRopeNode makeLeafRopeNode;
+
+        public MakeConcatNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        public abstract Rope executeMake(Rope left, Rope right, Encoding encoding);
+
+        @Specialization(guards = { "left.isEmpty()", "right.getEncoding() == encoding" })
+        public Rope concatEmptyLeftSameEncoding(Rope left, Rope right, Encoding encoding) {
+            return right;
+        }
+
+        @Specialization(guards = { "left.isEmpty()", "right.getEncoding() != encoding" })
+        public Rope concatEmptyLeftDifferentEncoding(Rope left, Rope right, Encoding encoding) {
+            return RopeOperations.withEncoding(right, encoding);
+        }
+
+        @Specialization(guards = { "right.isEmpty()", "left.getEncoding() == encoding" })
+        public Rope concatEmptyRightSameEncoding(Rope left, Rope right, Encoding encoding) {
+            return left;
+        }
+
+        @Specialization(guards = { "right.isEmpty()", "left.getEncoding() != encoding" })
+        public Rope concatEmptyRightDifferentEncoding(Rope left, Rope right, Encoding encoding) {
+            return RopeOperations.withEncoding(left, encoding);
+        }
+
+        @Specialization(guards = { "!left.isEmpty()", "!right.isEmpty()", "left.byteLength() < SHORT_LEAF_BYTESIZE_THRESHOLD", "right.byteLength() < SHORT_LEAF_BYTESIZE_THRESHOLD" })
+        public Rope concatLeaves(LeafRope left, LeafRope right, Encoding encoding,
+                                 @Cached("createBinaryProfile()") ConditionProfile sameCodeRangeProfile,
+                                 @Cached("createBinaryProfile()") ConditionProfile brokenCodeRangeProfile) {
+            if (makeLeafRopeNode == null) {
+                CompilerDirectives.transferToInterpreter();
+                makeLeafRopeNode = insert(RopeNodesFactory.MakeLeafRopeNodeGen.create(getContext(), getSourceSection(), null, null, null));
+            }
```

Comment:
That node could be in `@Cached` as well to make it more terse and avoid having a field for all instances of the node.